### PR TITLE
[Enhencement](Nereids) rules to optimize length function with aggregate function

### PR DESCRIPTION
**Pull request by LiBinfeng-01**
Add two rules in order to optimize plan in clickbench q27
- change expression string = "" to length(string) <>0 
- add length function in aggregate function agg(length) to child projection

example：
select t1.string from t1 where t1.string <> ""; ==>select t1.string from t1 where length(t1.string) > 0;
select avg(length(t1.string)) from t1; 
   from agg(length) to agg->project(length)

### CODE REVIEW

**Comment by LiBinfeng-01** [comment]
run buildall

**Comment by englefly** [comment]
it is better to do this transformation:
t1.string = ""
=>
length(t1.string)=0

It covers more cases.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/NotEqualToLength.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+            matchesTopType(Not.class).then(NotEqualToLength::rewrite)
+        );
+    }
+
+    private static Expression rewrite(Not not) {
+        Expression expr = not;
+        if (not.getArgument(0) instanceof EqualPredicate) {
+            EqualPredicate equalPredicate = (EqualPredicate) not.getArgument(0);
+            if (equalPredicate.getArgument(0).getDataType().isStringType()
+                    && equalPredicate.getArgument(1).equals(new StringLiteral(""))) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/NotEqualToLength.java:55]: how about varchar or char literal? should check StringLikeLiteral and getStringValue from it

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/NotEqualToLength.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+                expr = new GreaterThan(new Length(equalPredicate.getArgument(0)), new IntegerLiteral(0), false);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/NotEqualToLength.java:56]: use two arg ctor is enough

**Comment by LiBinfeng-01** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/NotEqualToLength.java (hunk 2 above)
Comment (LiBinfeng-01) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/NotEqualToLength.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/NotEqualToLength.java (hunk 1 above)
Comment (LiBinfeng-01) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/NotEqualToLength.java]: done

**Comment by LiBinfeng-01** [comment]
> it is better to do this transformation: t1.string = "" => length(t1.string)=0
> 
> It covers more cases.

have changed and add more cases abort that

**Comment by LiBinfeng-01** [comment]
run buildall

**Comment by LiBinfeng-01** [comment]
run buildall

**Review by englefly — DISMISSED** [review]

**Comment by LiBinfeng-01** [comment]
run buildall

**Comment by LiBinfeng-01** [comment]
run buildall

**Comment by py023** [comment]
run external
run feut

**Pull request closed without merging**